File: web/src/api/datasets.ts
import { readApiError } from './errors'
import type {
  DatasetCreateRequest,
  DatasetCreateResponse,
  DatasetDetailResponse,
  DatasetListPageResponse,
  DatasetStatusResponse,
  DatasetWorkflowErrorResponse,
} from '../types/datasets'

export async function createDataset(payload: DatasetCreateRequest): Promise<DatasetCreateResponse> {
  const response = await fetch('/api/datasets', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!response.ok) {
    throw new Error(await readApiError(response, 'Failed to create dataset'))
  }
  return response.json() as Promise<DatasetCreateResponse>
}

export async function fetchDatasets(): Promise<DatasetListPageResponse> {
  const response = await fetch('/api/datasets')
  if (!response.ok) {
    throw new Error(await readApiError(response, 'Failed to load datasets'))
  }
  return response.json() as Promise<DatasetListPageResponse>
}

export async function fetchDatasetDetail(datasetId: string): Promise<DatasetDetailResponse> {
  const response = await fetch(`/api/datasets/${datasetId}`)
  if (!response.ok) {
    throw new Error(await readApiError(response, 'Failed to load dataset detail'))
  }
  return response.json() as Promise<DatasetDetailResponse>
}

export async function fetchDatasetStatus(datasetId: string): Promise<DatasetStatusResponse> {
  const response = await fetch(`/api/datasets/${datasetId}/status`)
  if (!response.ok) {
    throw new Error(await readApiError(response, 'Failed to load dataset status'))
  }
  return response.json() as Promise<DatasetStatusResponse>
}

export async function retryDataset(datasetId: string): Promise<DatasetCreateResponse> {
  const response = await fetch(`/api/datasets/${datasetId}/retry`, {
    method: 'POST',
  })
  if (!response.ok) {
    throw new Error(await readApiError(response, 'Failed to retry dataset'))
  }
  return response.json() as Promise<DatasetCreateResponse>
}

export async function fetchDatasetWorkflowErrors(datasetId: string): Promise<DatasetWorkflowErrorResponse> {
  const response = await fetch(`/api/datasets/${datasetId}/workflow-errors`)
  if (!response.ok) {
    throw new Error(await readApiError(response, 'Failed to load dataset workflow errors'))
  }
  return response.json() as Promise<DatasetWorkflowErrorResponse>
}

export async function deleteDataset(datasetId: string): Promise<void> {
  const response = await fetch(`/api/datasets/${datasetId}`, {
    method: 'DELETE',
  })
  if (!response.ok) {
    throw new Error(await readApiError(response, 'Failed to delete dataset'))
  }
}

export async function downloadDatasetParquet(datasetId: string): Promise<void> {
  const response = await fetch(`/api/datasets/${datasetId}/download`)
  if (!response.ok) {
    throw new Error(await readApiError(response, 'Failed to download dataset parquet'))
  }

  const blob = await response.blob()
  const disposition = response.headers.get('content-disposition') ?? ''
  const filenameMatch = disposition.match(/filename="?([^"]+)"?/i)
  const filename = filenameMatch?.[1] ?? `${datasetId}.parquet`

  const url = window.URL.createObjectURL(blob)
  try {
    const link = document.createElement('a')
    link.href = url
    link.download = filename
    link.click()
  } finally {
    window.URL.revokeObjectURL(url)
  }
}

File: src/app/api/routes/datasets.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, status
from fastapi.responses import FileResponse

from app.api.deps import ApiDependencies, get_deps
from app.datasets import (
    DatasetCreateRequest,
    DatasetCreateResponse,
    DatasetDetailResponse,
    DatasetListPageResponse,
    DatasetStatusResponse,
    DatasetWorkflowErrorResponse,
)

router = APIRouter(prefix="/datasets", tags=["datasets"])


@router.post("", response_model=DatasetCreateResponse, status_code=status.HTTP_202_ACCEPTED)
def create_dataset(payload: DatasetCreateRequest, deps: ApiDependencies = Depends(get_deps)) -> DatasetCreateResponse:
    return deps.datasets.submit(payload)


@router.get("", response_model=DatasetListPageResponse)
def list_datasets(deps: ApiDependencies = Depends(get_deps)) -> DatasetListPageResponse:
    return deps.datasets.list_datasets()


@router.get("/{dataset_id}", response_model=DatasetDetailResponse)
def get_dataset(dataset_id: str, deps: ApiDependencies = Depends(get_deps)) -> DatasetDetailResponse:
    detail = deps.datasets.get_detail(dataset_id)
    if detail is None:
        raise HTTPException(status_code=404, detail=f"Dataset '{dataset_id}' not found")
    return detail


@router.get("/{dataset_id}/status", response_model=DatasetStatusResponse)
def get_dataset_status(dataset_id: str, deps: ApiDependencies = Depends(get_deps)) -> DatasetStatusResponse:
    status_payload = deps.datasets.get_status(dataset_id)
    if status_payload is None:
        raise HTTPException(status_code=404, detail=f"Dataset '{dataset_id}' not found")
    return status_payload


@router.post("/{dataset_id}/retry", response_model=DatasetCreateResponse, status_code=status.HTTP_202_ACCEPTED)
def retry_dataset(dataset_id: str, deps: ApiDependencies = Depends(get_deps)) -> DatasetCreateResponse:
    try:
        return deps.datasets.retry_dataset(dataset_id)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except RuntimeError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.get("/{dataset_id}/workflow-errors", response_model=DatasetWorkflowErrorResponse)
def get_dataset_workflow_errors(dataset_id: str, deps: ApiDependencies = Depends(get_deps)) -> DatasetWorkflowErrorResponse:
    error_payload = deps.datasets.get_workflow_errors(dataset_id)
    if error_payload is None:
        raise HTTPException(status_code=404, detail=f"Dataset '{dataset_id}' not found")
    return error_payload


@router.delete("/{dataset_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_dataset(dataset_id: str, deps: ApiDependencies = Depends(get_deps)) -> None:
    if not deps.datasets.delete_dataset(dataset_id):
        raise HTTPException(status_code=404, detail=f"Dataset '{dataset_id}' not found")


@router.get("/{dataset_id}/download")
def download_dataset(dataset_id: str, deps: ApiDependencies = Depends(get_deps)) -> FileResponse:
    path = deps.datasets.get_dataset_parquet_path(dataset_id)
    if path is None:
        raise HTTPException(status_code=404, detail=f"Dataset '{dataset_id}' parquet not found")
    return FileResponse(
        path,
        media_type="application/octet-stream",
        filename=path.name,
    )
